### DISCUSSION TITLE
DRM Preview Overlay

### DISCUSSION THREAD

**Discussion by ancient-geek** [raspberrypi/picamera2#407] [Ideas]
With the current drm_preview.py, overlays appear to be limited to the extent of the preview window.

However, there are use cases where it is helpful for the overlay plane to extend over the whole of the display whilst the preview is in a smaller window.  

For example, one might want to use the overlay to place GUI elements around the viewfinder window, or to display image statistics or camera configuration settings outside, rather than on top of, of the preview window.    

As a noob I've achieved this by changing line 170 from:
   `self.crtc.set_plane(self.overlay_plane, self.overlay_fb, x, y, w, h, 0, 0, width, height)`
to
   `self.crtc.set_plane(self.overlay_plane, self.overlay_fb, x, y, width, height, 0, 0, width, height)`

However, perhaps this might be a generally worthwhile approach as it gives the user a choice of overlay size?

**Comment by davidplowman** [comment]
Hi, an interesting suggestion. Certainly if this works for your application, then go with it!

More generally, I'm a bit conflicted about the whole thing. On the one hand it seems like a good idea to do something a bit more generic like this. One might ask: should we still allow resizing of the overlay (though to a different size)? Does that make the DRM preview more incompatible with the QT or QTGL ones? Or perhaps we don't care? One thing I really want to avoid is devoting time to beefing up all the different preview implementations in various ways, though perhaps there's more justification in the DRM case because there's no GUI framework. I'm not sure. I'm certainly open to opinions and persuasion on this one...

**Reply by ancient-geek** [reply]
I do agree that the lack of a GUI distinguishes DRM from QT or QTGL and that being able to use the picamera2 overlay to implement a basic GUI around the viewfinder seems attractive to me. For example, I've been able to implement this: 

![peica](https://user-images.githubusercontent.com/5341224/200939806-1b7b073b-bddb-49ef-820c-43a8f500ba1d.jpg)

On the other hand, I appreciate that diverging the DRM preview from QT/QTGL would add complexity. 
Given that modification of a single line of picamera2 code appears to do the trick, perhaps it would be simplest to leave picamera2 as is and just point out this possibility somewhere in the documentation?

**Reply by davidplowman** [reply]
Thanks for posting the picture, looks great! I guess I'm vacillating on this one, maybe I just need a little time for my view of the world to be adjusted and then adding a feature to the DRM preview that the others don't need would seem OK...

**Comment by s12wu** [comment]
For anyone trying this after commit https://github.com/raspberrypi/picamera2/commit/50e78bfa404d0754abad632ec93df66e4af77334

The line  
`ctx.add_plane(self.overlay_plane, self.overlay_fb, self.crtc, (0, 0, width, height), (x, y, w, h))`  
now needs to be replaced with
`ctx.add_plane(self.overlay_plane, self.overlay_fb, self.crtc, (0, 0, width, height), (x, y, width, height))`
to achieve the same effect

Or, if you want the overlay span the entire display with the viewfinder preview not at the top left (0, 0)
`ctx.add_plane(self.overlay_plane, self.overlay_fb, self.crtc, (0, 0, width, height), (0, 0, width, height))`

(I don't know if @ancient-geek's original change did this)
